cloudinit/ssh_util.py

```diff
@@ -270,7 +270,6 @@ def setup_user_keys(keys, username, options=None):
     (auth_key_fn, auth_key_entries) = extract_authorized_keys(username)
     with util.SeLinuxGuard(ssh_dir, recursive=True):
         content = update_authorized_keys(auth_key_entries, key_entries)
-        util.ensure_dir(os.path.dirname(auth_key_fn), mode=0o700)
```

Comment:
I'm certainly interested in supporting this use-case "global ssh key folder";  where we left off with this was in the bug:

> Right, sshd will assume the user's uid before reading the authorized_keys file.
> This means each directory in the path to the key file must be readable by the
> target user.
> 
> That generally means for the path, we need to visit each layer and verify
> that the user has access to each path.
> 

Removing "ensure_dir" here works for your scenario where the image has an
existing directory with correct permissions; however removing it will not work
for other cases; in particular, if the directory cloud-init writes keys to does exist
and is not 0700 and is not *meant* to be readable by anyone, then we're exposing
keys when we would not.

To move this user-story forward, we're likely going to need some new ssh_authorized_keys config options, something like:

```
ssh_authorized_keys:
  - ssh-rsa AAAAA4
  - ssh-rsa AAAAB3
ssh_authorized_keys_permissions:
   directory: 0755
   keys: 0644
   owner: root
```

something like that will let this code default to the current settings
but let users override the values to ensure_dir, write_file and
chownbyid.

I believe the above values I proposed should enable this system-wide
use-case.

Reply:
hi @raharper 
>however removing it will not work
for other cases; in particular, if the directory cloud-init writes keys to does exist
and is not 0700 and is not meant to be readable by anyone, then we're exposing
keys when we would not.

Other cases are ssh defaults, which are covered by https://github.com/canonical/cloud-init/blob/master/cloudinit/ssh_util.py#L260. 
If some non-default sshd configurations are used cloud-init shouldn't just guess a right permissions for unknown settings.
If default sshd setting were modified by user he also responsible to modify corresponding system settings to match sshd modification.
`700` is critical permission for `~/.ssh` only because this folder **by default** stores **private** ssh keys, authorized_keys and public ssh keys are not sensitive information.

Reply:
> `700` is critical permission for `~/.ssh` only because this folder **by default** stores **private** ssh keys, authorized_keys and public ssh keys are not sensitive information.

How is cloud-init supposed to know whether or not the contents of the directory it's writing to has private data or not?  We cannot know unless the user-data explicitly says, please mark this authorized_keys_dir with a specific permission set.  This is what I proposed previously with new cloud-config as above.  This absolves cloud-init from having to guess whether or not the target location is safe.

Reply:
I completely got your idea regarding additional option `ssh_authorized_keys_permissions` for cloud-init but it will only add a complexity to the code without any value to the functionality.

I'll explain why:
Permissions can be safely set _only_ during resource creation. This called a **default**. When default was set there are two options:

1. Resource wasn't changed externally and it still has a default permissions. Nothing to do in this case.
2. Resource was changed externally by user. Changing defaults immediately transfers responsibility from application to user. If application will roll back this changes this will be called a **forcing**.

Even if forcing looks reasonable from developers perspective, in unix world _only user_ can decide which configuration he wants/needs. The way which uses forcing is wrong by itself. For example think about it from this perspective:

- `chmod -R 777 /etc` should kernel rollback this changes because system is smarter then user? Or `chmod` should have an hard-coded paths which it will ignore because user who can trigger that command is not smart enough?
- Shall we also rollaback changes to `/etc/ssh/sshd_config` maybe? Because user can potentially allow root access even without password?

The only person who can decide to use enforcing or not is user of the system and he already has `bootcmd` and`runcmd` for this. We do not need options `ssh_authorized_keys_permissions`, `etc_shadow_permissions`, `var_log_permissions`, etc.

Reply:
cloud-init is certainly opinionated; so for sure we force certain things.  We delete ssh host keys by default on new instances even if someone were to have pre-generated them.  You can specify config to *disable* the default behavior that we are forcing; as in general, host keys in images aren't there because someone put them there; it's more likely someone captured an instance and tried to make a template.  

That aside;  Instead of removing ensure dir entirely, if we only run ensure_dir if the target directory does not exit; this supports both our use-cases.

```
diff --git a/cloudinit/ssh_util.py b/cloudinit/ssh_util.py
index c3a9b5b..0d68135 100644
--- a/cloudinit/ssh_util.py
+++ b/cloudinit/ssh_util.py
@@ -268,9 +268,11 @@ def setup_user_keys(keys, username, options=None):
 
     # Extract the old and make the new
     (auth_key_fn, auth_key_entries) = extract_authorized_keys(username)
+    key_dir = os.path.dirname(auth_key_fn)
     with util.SeLinuxGuard(ssh_dir, recursive=True):
         content = update_authorized_keys(auth_key_entries, key_entries)
-        util.ensure_dir(os.path.dirname(auth_key_fn), mode=0o700)
+        if not os.path.exists(key_dir):
+            util.ensure_dir(key_dir, mode=0o700)
         util.write_file(auth_key_fn, content, mode=0o600)
         util.chownbyid(auth_key_fn, pwent.pw_uid, pwent.pw_gid)
```